packages/data-service/src/data-service.ts

```diff
@@ -2285,6 +2306,44 @@ class DataServiceImpl extends WithLogContext implements DataService {
     return { name, ...normalized };
   }
 
+  @op(mongoLogId(1_001_000_063))
+  async previewUpdate(
+    ns: string,
+    filter: Document,
+    update: Document
+  ): Promise<UpdatePreview> {
+    const coll = this._collection(ns, 'CRUD');
+    const session = this._startSession('CRUD');
+    try {
+      session.startTransaction();
+
+      const docsToPreview = await coll
+        .find(filter, { session })
+        .sort({ _id: 1 })
+        .limit(10)
+        .toArray();
+      const idsToPreview = docsToPreview.map((doc) => doc._id);
+      await coll.updateMany({ _id: { $in: idsToPreview } }, update, {
+        session,
+      });
+      const changedDocs = await coll
+        .find({ _id: { $in: idsToPreview } }, { session })
+        .sort({ _id: 1 })
+        .toArray();
+
+      const changes = docsToPreview.map((before, idx) => ({
+        before,
+        after: changedDocs[idx],
+      }));
+      return { changes, serverError: undefined };
```

Comment:
It depends on the conventions we already have. I think having the error here is more explicit in case it's the user fault (like breaking an index constraint) and bubble up on other errors (we could have a mapping function that does that).

Reply:
It is just that _usually_ errors get thrown and results returned. Then the calling code can catch the error and deal with it. You can't have both a result AND an error anyway.

Reply:
Nit: i would also prefer if we keep it consistent, if throwing is what we do everywhere else, can we keep it the same? @kmruiz would there be a downside in doing so?

Reply:
OK I'll let the errors bubble up, but if it's ok for you, I'll keep the wrapper object, just in case it needs to evolve later.

Reply:
@mcasimir Didn't see your comment, sorry 😄 .

I think in the data service everything throws an exception, so you are right, let's keep it consistent.